### PR TITLE
Refactor forms with `#build_from_application` and `id` as a parameter

### PR DESCRIPTION
**Pull request by chubberlisk**
## Context

Master failed (https://dfe-ssp.visualstudio.com/Become-A-Teacher/_build/results?buildId=37643&view=logs&j=32f3c9c7-a610-55d2-616a-efba1ecc24a2&t=85c1dfa3-8bc2-5353-acab-1c1a0e7b38db) because of:

```
  1) CandidateInterface::OtherQualificationForm.build_all_from_application orders other qualifications by created at
     Failure/Error:
       form.application_qualifications.create(
         id: 1,
         level: 'other',
         qualification_type: 'BTEC',
         subject: 'Being a Superhero',
         institution_name: 'School of Heroes',
         grade: 'Distinction',
         predicted_grade: false,
         award_year: '2012',
         created_at: DateTime.new(2019, 1, 1, 21, 0, 0),

     ActiveRecord::RecordNotUnique:
       PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "application_qualifications_pkey"
       DETAIL:  Key (id)=(1) already exists.
```

This is a flaky test and related to https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/869. Turns out that the `OtherQualificationForm` doesn't need to take in an `id` and can instead take in an `application_qualification` object so that finding it can be handled by the controller. This means that tests aren't needed for asserting `id`s and that these form objects are consistent with other ones like work history and GCSEs.

There are a few other form objects that take in an `application_form` and some form of `id` so this PR refactors those as well.

## Changes proposed in this pull request

This PR refactors:

- `OtherQualificationForm#build_from_application` to be `#build_from_qualification`
- `DegreeForm#build_from_application` to be `#build_from_qualification`
- `VolunteeringRoleForm#build_from_application` to be `#build_from_experience`

## Guidance to review

- Review commit by commit
- Does this make sense?
- Have any other forms been missed?

## Link to Trello card

N/A

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

### CODE REVIEW

File: spec/models/candidate_interface/other_qualification_form_spec.rb [hunk 1]
```diff
@@ -107,23 +106,20 @@
...
-          qualification_type: 'BTEC',
-          subject: 'Being a Sidekick',
-          institution_name: 'School of Sidekicks',
-          grade: 'Merit',
-          predicted_grade: false,
-          award_year: '2010',
-        )
-        form.application_qualifications.create(id: 2, level: 'other')
-      end
+  describe '.build_from_qualification' do
```
Comment (duncanjbrown) [inline spec/models/candidate_interface/other_qualification_form_spec.rb]: :+1: 

There's another `id` left in this file which I think we could squash in another commit. On line 159 we test the update action by creating some post data including an ID, then creating an application qualification with the same id. This is probably liable to the same problems with IDs that we saw breaking in `master`. A fix might be to create the existing qualification first, then use the id in the post data.

Before

```ruby
form_data = {
  id: 1,
  qualification_type: 'BTEC',
  subject: 'Being a Everyday Hero',
  institution_name: 'School of Humans',
  grade: 'Distinction',
  award_year: '2011',
}
qualification = CandidateInterface::OtherQualificationForm.new(form_data)
application_form = create(:application_form) do |form|
  form.application_qualifications.create(
    id: 1,
    level: 'other',
    qualification_type: 'BTEC',
    subject: 'Being a Everyday Hero',
    institution_name: 'School of Hoomans',
    grade: 'Pass',
    predicted_grade: false,
    award_year: '2011',
  )
end

expect(qualification.update(application_form)).to eq(true)
expect(application_form.application_qualifications.other.first)
  .to have_attributes(form_data)
```

After 
```ruby
application_form = create(:application_form)

existing_qualification = application_form.application_qualifications.create(
  level: 'other',
  qualification_type: 'BTEC',
  subject: 'Being a Everyday Hero',
  institution_name: 'School of Hoomans',
  grade: 'Pass',
  predicted_grade: false,
  award_year: '2011',
)

form_data = {
  id: existing_qualification.id,
  qualification_type: 'BTEC',
  subject: 'Being a Everyday Hero',
  institution_name: 'School of Humans',
  grade: 'Distinction',
  award_year: '2011',
}
qualification_form = CandidateInterface::OtherQualificationForm.new(form_data)

expect(qualification_form.update(application_form)).to eq(true)
expect(application_form.application_qualifications.other.first)
  .to have_attributes(form_data)
```

File: spec/models/candidate_interface/degree_form_spec.rb [hunk 2]
```diff
@@ -78,7 +78,6 @@
     it 'creates an array of objects based on the provided ApplicationForm' do
```
Comment (duncanjbrown) [inline spec/models/candidate_interface/degree_form_spec.rb]: We could dispose of the explicit ID at line 266 in a similar way — might need to fix up the lets a bit

File: spec/models/candidate_interface/volunteering_role_form_spec.rb [hunk 3]
```diff
@@ -28,9 +28,8 @@
   describe '.build_all_from_application' do
```
Comment (duncanjbrown) [inline spec/models/candidate_interface/volunteering_role_form_spec.rb]: Ditto for the `update` spec in this file (line 95)

File: spec/components/work_history_review_component_spec.rb [hunk 4]
```diff
@@ -14,17 +14,14 @@
   context 'when there is not a break in the work history' do
     let(:application_form) do
       create(:application_form) do |form|
-        data[:id] = 1
```
Comment (duncanjbrown) [inline spec/components/work_history_review_component_spec.rb]: Noice!

File: spec/models/candidate_interface/other_qualification_form_spec.rb (hunk 1 above)
Comment (chubberlisk) [reply spec/models/candidate_interface/other_qualification_form_spec.rb]: This fix makes sense to me. Thank you!

Comment (chubberlisk) [reply spec/models/candidate_interface/other_qualification_form_spec.rb]: Done in the commit about refactoring the `OtherQualificationForm` and spec.

File: spec/models/candidate_interface/degree_form_spec.rb (hunk 2 above)
Comment (chubberlisk) [reply spec/models/candidate_interface/degree_form_spec.rb]: Done in the commit about refactoring the `DegreeForm` and spec.

File: spec/models/candidate_interface/volunteering_role_form_spec.rb (hunk 3 above)
Comment (chubberlisk) [reply spec/models/candidate_interface/volunteering_role_form_spec.rb]: Done in the commit about refactoring the `VolunteeringRoleForm` and spec.

**Pull request merged**